src/riak_kv_compile_tab.erl

```diff
@@ -71,35 +100,75 @@ is_compiling(BucketType) ->
         compiling_state() | notfound.
 get_state(BucketType) when is_binary(BucketType) ->
     case dets:lookup(?TABLE, BucketType) of
-        [{_,_,_,State}] ->
+        [{_,_,_,_,State}] ->
             State;
         [] ->
             notfound
     end.
 
 %%
+-spec get_compiled_ddl_version(BucketType :: binary()) ->
+    riak_ql_ddl:compiler_version_type() | notfound.
+get_compiled_ddl_version(BucketType) when is_binary(BucketType) ->
+    case dets:match(?TABLE, {BucketType,'$1','_','_','_'}) of
+        [[Version]] ->
+            Version;
+        [] ->
+            notfound
+    end.
+
+%%
 -spec get_ddl(BucketType :: binary()) ->
         term() | notfound.
 get_ddl(BucketType) when is_binary(BucketType) ->
     case dets:lookup(?TABLE, BucketType) of
-        [{_,DDL,_,_}] ->
+        [{_,_,DDL,_,_}] ->
             DDL;
         [] ->
             notfound
     end.
 
 %% Update the compilation state using the compiler pid as a key.
+%% Since it has an active Pid, it is assumed to have a DDL version already.
 -spec update_state(CompilerPid :: pid(), State :: compiling_state()) ->
         ok | notfound.
 update_state(CompilerPid, State) when is_pid(CompilerPid),
                                        ?is_compiling_state(State) ->
-    case dets:match(?TABLE, {'$1','$2',CompilerPid,'_'}) of
-        [[BucketType, DDL]] ->
-            insert(BucketType, DDL, CompilerPid, State);
+    case dets:match(?TABLE, {'$1','$2','$3',CompilerPid,'_'}) of
+        [[BucketType, DDLVersion, DDL]] ->
+            insert(BucketType, DDLVersion, DDL, CompilerPid, State);
         [] ->
             notfound
     end.
 
+%% Mark any lingering compilations as being retried
+-spec mark_compiling_for_retry() -> ok.
+mark_compiling_for_retry() ->
+    CompilingPids = dets:match(?TABLE, {'_','_','_','$1',compiling}),
+    lists:foreach(fun([Pid]) -> update_state(Pid, retrying) end, CompilingPids).
+
+%% Get the list of tables which need to be recompiled
+-spec get_tables_needing_recompiling(DDLVersion :: riak_ql_ddl:compiler_version_type()) ->
```

Comment:
`records` seems both self-evident and less awkward, but perhaps I'm missing the ambiguity there.